bpf/utils.h

```diff
@@ -0,0 +1,278 @@
+    } else {
+        // TODO : Need to implement other specific ethertypes if needed
+        // For now other parts of flow id remain zero
+        __builtin_memset(&(id->src_ip), 0, sizeof(struct in6_addr));
+        __builtin_memset(&(id->dst_ip), 0, sizeof(struct in6_addr));
+        id->transport_protocol = 0;
+        id->src_port = 0;
+        id->dst_port = 0;
+    }
+    return SUBMIT;
+}
+
+static inline void set_key_with_l2_info(struct sk_buff *skb, u16 eth_protocol, flow_id *id) {
+     struct ethhdr eth;
+     __builtin_memset(&eth, 0, sizeof(eth));
+     bpf_probe_read(&eth, sizeof(eth), (struct ethhdr *)(skb->head + skb->mac_header));
+     id->eth_protocol = eth_protocol;
+     __builtin_memcpy(id->dst_mac, eth.h_dest, ETH_ALEN);
+     __builtin_memcpy(id->src_mac, eth.h_source, ETH_ALEN);
+ }
+
+static inline void set_key_with_l3_info(struct sk_buff *skb, u16 family, flow_id *id, u8 *protocol) {
+     if (family == AF_INET) {
+         struct iphdr ip;
+         __builtin_memset(&ip, 0, sizeof(ip));
+         bpf_probe_read(&ip, sizeof(ip), (struct iphdr *)(skb->head + skb->network_header));
+         __builtin_memcpy(id->src_ip, ip4in6, sizeof(ip4in6));
+         __builtin_memcpy(id->dst_ip, ip4in6, sizeof(ip4in6));
+         __builtin_memcpy(id->src_ip + sizeof(ip4in6), &ip.saddr, sizeof(ip.saddr));
+         __builtin_memcpy(id->dst_ip + sizeof(ip4in6), &ip.daddr, sizeof(ip.daddr));
+         *protocol = ip.protocol;
+     } else if (family == AF_INET6) {
+         struct ipv6hdr ip;
+         __builtin_memset(&ip, 0, sizeof(ip));
+         bpf_probe_read(&ip, sizeof(ip), (struct ipv6hdr *)(skb->head + skb->network_header));
+         __builtin_memcpy(id->src_ip, ip.saddr.in6_u.u6_addr8, IP_MAX_LEN);
+         __builtin_memcpy(id->dst_ip, ip.daddr.in6_u.u6_addr8, IP_MAX_LEN);
+         *protocol = ip.nexthdr;
+     }
+ }
+
+static inline void set_key_with_l4_info(struct sk_buff *skb, flow_id *id, u8 protocol, u16 *flags) {
+     u16 sport = 0,dport = 0;
+     struct tcphdr tcp;
+
+     __builtin_memset(&tcp, 0, sizeof(tcp));
+     bpf_probe_read(&tcp, sizeof(tcp), (struct tcphdr *)(skb->head + skb->transport_header));
+     sport = bpf_ntohs(tcp.source);
+     dport = bpf_ntohs(tcp.dest);
+     set_flags(&tcp, flags);
+     id->src_port = sport;
+     id->dst_port = dport;
+     id->transport_protocol = protocol;
+ }
+
+static inline long tcp_drop_lookup_and_update_flow(struct sk_buff *skb, flow_id *id, u8 state, u16 flags,
+                                            enum skb_drop_reason reason) {
+     flow_metrics *aggregate_flow = bpf_map_lookup_elem(&aggregated_flows, id);
+     if (aggregate_flow != NULL) {
+         aggregate_flow->tcp_drops.packets += 1;
+         aggregate_flow->tcp_drops.bytes += skb->len;
+         aggregate_flow->tcp_drops.state = state;
+         aggregate_flow->tcp_drops.flags = flags;
+         aggregate_flow->tcp_drops.drop_cause = reason;
```

Comment:
is bitwise OR possible for these, to accumulate values? I think at least for flags we can do it.
Also I wonder if we could prioritize drop reasons, e.g. if there can be unknown reasons vs known reason we would privilege the known one? does that make sense?

Reply:
we already filter out unknonwn reasons before we come to this function, regarding the flags yeah I think we can or

Reply:
but like we do `aggregate_flow->tcp_drops.bytes += skb->len`, which means basically we add the input data into the aggregate object, why don't we do the same for flags: `aggregate_flow->tcp_drops.flags |= flags;` (instead of just `=`) ?

Reply:
will fix it